heat/core/operations.py

```diff
@@ -193,33 +190,43 @@ def argmin(x, axis=None):
     Examples:
     --------
 
-    >>> a = ht.randn(3,3)
+    >>> a = ht.random.randn(3,3)
     >>> a
-    tensor([[-1.7297,  0.2541, -0.1044],
-            [ 1.0865, -0.4415,  1.3716],
-            [-0.0827,  1.0215, -2.0176]])
+    tensor([[ 0.2604, -0.2292,  1.2846],
+        [-0.6166,  0.4549, -0.1443],
+        [-1.0306, -1.1125,  0.5686]])
     >>> ht.argmin(a)
-    tensor([8])
+    tensor([7])
     >>> ht.argmin(a, axis=0)
-    tensor([[0, 1, 2]])
+    tensor([[2, 2, 1]])
     >>> ht.argmin(a, axis=1)
-    tensor([[0],
-            [1],
-            [2]])
-    '''
+    tensor([[1],
+        [0],
+        [1]])
+    >>> out = ht.zeros((1, 3))
+    >>> ht.argmin(a, axis=0, out=out)
+    >>> out
+    tensor([[2, 2, 1]])
 
+    '''
     if axis is None:
-        # TEMPORARY SOLUTION! TODO: implementation for axis=None, distributed tensor Issue #100
-        # perform sanitation
-        if not isinstance(x, tensor.tensor):
-            raise TypeError(
-                'expected x to be a ht.tensor, but was {}'.format(type(x)))
-        axis = stride_tricks.sanitize_axis(x.shape, axis)
-        out = torch.reshape(torch.argmin(x._tensor__array), (1,))
-        return tensor.tensor(out, out.shape, types.canonical_heat_type(out.dtype), split=None, comm=x.comm)
+        # flatten tensor
+        x_flat = torch.reshape(x._tensor__array, (x._tensor__array.numel(),))
+        # apply torch.min to x_flat with axis=0, keep indices only (second element)
+        result = __reduce_op(tensor.tensor(x_flat, x_flat.shape, types.canonical_heat_type(x.dtype), x.split, x.comm), torch.min, MPI.MIN,
+                             axis=0, out=None)._tensor__array[1]
+    else:
+        # argmin() is the second element of torch.min()
+        result = __reduce_op(x, torch.min, MPI.MIN, axis,
```

Comment:
> The usage of MPI.MIN is not correct. It should rather be MPI.MINLOC as we are trying to realize argmin(). Just computing the minimum of the indices is not sufficient, as I may have a split tensor a such:
> 
> Node 0:
> [4, 5, 6] - argmin() -> [0] (index)
> 
> Node 1:
> [3, 2, 1] - argmin() -> [2] (index)
> 
> As in code:
> min([0], [2 + 3 (global offset)]) -> [0]
> 
> Intended behavior:
> argmin([4, 5, 6, 3, 2, 1]) -> [5]
> 
> Also, the global offset of the split tensor parts are not considered in the above code.

Hi Markus, no, that's not how it works. At the moment ht.argmin()  returns  the second value of torch.min(), just as ht.min() returns the first value of torch.min(). There is no computing the minimum of the indices anywhere at the moment. 

I do agree that MPI.MINLOC in combination with torch.argmin is probably a better solution. I'll look into it.

Reply:
Using torch.argmin()'s second return value is okay to obtain the local partial result. You will receive the minimal indices on the local device. 

Afterwards, the global result needs to be obtained (in case of a split). As far as I understand the code correctly, this is done in line 218-220 where the MPI.MIN reduction is performed. As it stands, this will compute the minimum of the partial indices.

Reply:
No, it computes the minimum of the flattened data (x_flat), and that result comes with the associated index. In fact I could replace:

```
if axis is None:
        # flatten tensor
        x_flat = torch.reshape(x._tensor__array, (x._tensor__array.numel(),))
        # apply torch.min to x_flat with axis=0, keep indices only (second element)
        result = __reduce_op(tensor.tensor(x_flat, x_flat.shape,   types.canonical_heat_type(x.dtype), x.split, x.comm), torch.min, MPI.MIN,
                             axis=0, out=None)._tensor__array[1]

``` 

with:

```
if axis is None:
        # flatten tensor
        x_flat = torch.reshape(x._tensor__array, (x._tensor__array.numel(),))
        # apply torch.min to x_flat with axis=0, keep indices only (second element)
        result = operations.min(x_flat, axis=0)._tensor__array[1]  #<-- second element only! 

``` 

Time for a phone call maybe?